src/boot/efi/boot.c

```diff
@@ -1722,10 +1732,18 @@ static int boot_entry_compare(const BootEntry *a, const BootEntry *b) {
         /* Now order by ID. The version is likely part of the ID, thus note that this will generatelly put
          * the newer versions earlier. Specifying a sort key explicitly is preferable, because it gives an
          * explicit sort order. */
-        r = -strverscmp_improved(a->id, b->id);
+        r = -strverscmp_improved(a->id_without_profile ?: a->id, b->id_without_profile ?: b->id);
         if (r != 0)
                 return r;
 
+        /* Let's sort profiles by their profile */
+        if (a->id_without_profile && b->id_without_profile &&
+            streq(a->id_without_profile, b->id_without_profile)) {
```

Comment:
If both `a` and `b` have `id_without_profile`, then they are equivalent in the sense of `strvserscmp_improved()`.
So, is `streq()` really necessary here?? If so, please add a short comment about that.